### PR TITLE
Allow user env vars for GitLab subgroups

### PR DESCRIPTION
**Pull request by geropl**
## Description

This PR allows to specify environment variable patterns for subgroups in GitLab. This is a long-standing issue that we always avoided to fix, but is now requested/assumed to be working by multiple customers.

We had multiple discussions in the past about deprecating this feature in the first place. However never did so, because it's a very convenient feature, that even if we had better secret support, adds value for users.

For examples of what's possible see the tests [here](https://github.com/gitpod-io/gitpod/pull/17831/files#diff-bec292cdaefd0cda4dcc4402f0b610f28a9a81d5d4a2f90c57e8abb73abcc4a7) and [here](https://github.com/gitpod-io/gitpod/blob/996620b21db10109f15f0598d90f659e9449db53/components/gitpod-protocol/src/user-env-var.spec.ts#L24-L145).

The docs don't need to be changed, but could use a bit more explanation (see [here](https://www.gitpod.io/docs/configure/projects/environment-variables#using-the-account-settings)).

## Related Issue(s)

Fixes WEB-353

### Resolution
 - add these user-level env vars to your user config: [link](https://github.com/gitpod-io/gitpod/blob/80b02f61afa159f126346e27863c6c74392dd2e7/components/server/src/workspace/env-var-service.spec.ts#L302-L350)
 - open two workspaces:
   - [project 1](https://gpl-gitlab-env-var.preview.gitpod-dev.com/#https://gitlab.com/geropl-test-group/subgroup1/test-project-1)
   - [project 2](https://gpl-gitlab-env-var.preview.gitpod-dev.com/#https://gitlab.com/geropl-test-group/test-project-2)
 - note that you only see the expected env vars only :heavy_check_mark: 

### Permissions 1
 - in one of the workspacex run `gp env -u USER_GROUP_TEST`, and see the "cannot delete" error :heavy_check_mark: 

### Permissions 2
- in one of the workspaces, run:
```
gp env TEST_NEW=abc
gp env
gp env -u TEST_NEW
gp env
```
and see that it does what you'd expect without errors :heavy_check_mark: 

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - gpl-env-var-1</li>
	<li><b>🔗 URL</b> - <a href="https://gpl-env-var-1.preview.gitpod-dev.com/workspaces" target="_blank">gpl-env-var-1.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [x] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
</details>
 
/hold

### CODE REVIEW

File: components/gitpod-protocol/src/protocol.ts [hunk 1]
```diff
@@ -466,12 +475,19 @@ export namespace UserEnvVar {
...
+    export function splitRepositoryPattern(pattern: string): string[] {
+        return pattern.split(DELIMITER);
+    }
+
+    function joinRepositoryPattern(...parts: string[]): string {
+        return parts.join(DELIMITER);
+    }
+
+    // DEPRECATED
     export function score(value: UserEnvVarValue): number {
```
Comment (svenefftinge) [inline components/gitpod-protocol/src/protocol.ts:491]: This doesn't seem to be used anymore

Comment (geropl) [reply components/gitpod-protocol/src/protocol.ts:491]: Ah, rebase artifact. :+1:

File: components/gitpod-protocol/src/protocol.ts [hunk 2]
```diff
@@ -452,9 +461,9 @@ export namespace UserEnvVar {
         if (pattern.trim() === "") {
             return "Scope must not be empty.";
         }
-        const split = pattern.split("/");
-        if (split.length < 2) {
-            return "A scope must use the form 'organization/repo'.";
+        const split = splitRepositoryPattern(pattern);
+        if (split.length < MIN_PATTERN_SEGMENTS) {
+            return "A scope must use the form 'organization/repo(/...)'.";
```
Comment (svenefftinge) [inline components/gitpod-protocol/src/protocol.ts:466]: I don't think the additional `(/...)` is helpful, given we use a very concrete example. I.e. there cannot be anything after the repo name.

File: components/gitpod-protocol/src/protocol.ts [hunk 3]
```diff
@@ -427,6 +427,15 @@ export interface UserEnvVar extends UserEnvVarValue {
 }
 
 export namespace UserEnvVar {
+    export const DELIMITER = "/";
+    export const WILDCARD_ASTERISK = "*";
+    const WILDCARD_SHARP = "#"; // TODO(gpl) Where does this come from? Bc we have/had patterns as part of URLs somewhere, maybe...?
```
Comment (svenefftinge) [inline components/gitpod-protocol/src/protocol.ts:432]: Where does this come from?

Comment (svenefftinge) [reply components/gitpod-protocol/src/protocol.ts:432]: there is a mention below https://github.dev/gitpod-io/gitpod/blob/2c1d74d479c10c68081d62c5a5f566ad1acd4e6d/components/gitpod-protocol/src/protocol.ts#L497-L498 not sure if this is still used and how/why a user would pass such a pattern through URL

File: components/gitpod-protocol/src/protocol.ts [hunk 4]
```diff
@@ -466,12 +475,19 @@ export namespace UserEnvVar {
         return undefined;
     }
 
-    // DEPRECATED: Use ProjectEnvVar instead of repositoryPattern - https://github.com/gitpod-com/gitpod/issues/5322
     export function normalizeRepoPattern(pattern: string) {
         return pattern.toLocaleLowerCase();
     }
 
-    // DEPRECATED: Use ProjectEnvVar instead of repositoryPattern - https://github.com/gitpod-com/gitpod/issues/5322
+    export function splitRepositoryPattern(pattern: string): string[] {
```
Comment (svenefftinge) [inline components/gitpod-protocol/src/protocol.ts:482]: no need to export

**Comment by svenefftinge** [comment]
It seems a bit surprising that a single `*` would match any number of segments. I believe it would be better to allow users being explicit about that using a globstar. Did you consider that? We could even just use a [glob library](https://github.com/isaacs/node-glob) which would also make it easier to explain users how the pattern match.

File: components/gitpod-protocol/src/protocol.ts (hunk 3 above)
Comment (geropl) [reply components/gitpod-protocol/src/protocol.ts:432]: It's from here: https://github.com/gitpod-io/gitpod/pull/17831/files#diff-957d9dd15b947e3969539ebd0dabcb8b40ce8b5564e5820cd9bb27d88be1dbc7L501

Could not find a mention in docs or tests anywhere. :shrug:
I*'m inclined to remove it, but want to do that in a separate PR.

Comment (svenefftinge) [reply components/gitpod-protocol/src/protocol.ts:432]: In that case I'd say lat's remove and see who complains 😅

Comment (geropl) [reply components/gitpod-protocol/src/protocol.ts:432]: Will do, after this has landed :+1:

File: components/gitpod-protocol/src/protocol.ts [hunk 5]
```diff
@@ -466,55 +475,92 @@ export namespace UserEnvVar {
...
+        matchedEnvVars.forEach((e) => {
             const l = resmap.get(e.name) || [];
             l.push(e);
             resmap.set(e.name, l);
         });
 
-        result = [];
+        // In cases of multiple matches per env var: find the best match
+        // Best match is the most specific pattern, where the left-most segment is most significant
+        function scoreMatchedEnvVar(e: T): number {
```
Comment (svenefftinge) [inline components/gitpod-protocol/src/protocol.ts:553]: nice simplification! Comparing `repositoryPattern.indexOf('*')` would be even simpler. Also I'd consider this just sorting function so maybe drop the `score`term?

Comment (geropl) [reply components/gitpod-protocol/src/protocol.ts:553]: > repositoryPattern.indexOf('*')

But that doesn't work for `x/*/z` vs `x/*/*` right? :thinking:

**Comment by geropl** [comment]
> We could even just use a [glob library](https://github.com/isaacs/node-glob) which would also make it easier to explain users how the pattern match.

Yes, considered that...

> Edit: going with a lib might be a bit too much,

...but also felt it's a tad too much as well. :upside_down_face: 

> but the supporting and requiring ** to match multiple segments would be good.

Hm, so far I only thought about `*/*` (required) vs `*` (does not work).
I like the idea, and will see how it pans out after lunch.

File: components/gitpod-protocol/src/protocol.ts (hunk 5 above)
Comment (svenefftinge) [reply components/gitpod-protocol/src/protocol.ts:553]: Oh, true. Also there is not always a pattern more specific than another. E.g. `x/*/z`is not more or less specific than `x/y/*`.
Thinking about it I'd say the comparing the number of wildcard segments would be most appropriate. wdyt?

**Comment by geropl** [comment]
@svenefftinge I added the `**` special case in [996620b](https://github.com/gitpod-io/gitpod/pull/17831/commits/996620b21db10109f15f0598d90f659e9449db53) and would love your feedback on the complexity.
Good news: it helped uncover a bug due to the additional tests :upside_down_face:

**Comment by svenefftinge** [comment]
I believe translating the pattern to a regexp (i.e. '**' becomes '.*' and '*' becomes '[^/]*') would be simpler.

**Comment by geropl** [comment]
@svenefftinge I did the regex version in [61f84cb](https://github.com/gitpod-io/gitpod/pull/17831/commits/61f84cb2b566dbb72d187fb0cec6aea071786582) but it a) does not remove that much code, and b) has other problems (escaping the pattern for other regex control characters `.+`, etc.).
As we have the other version here already and it's tested, I would like to go with that. Ok?

File: components/gitpod-protocol/src/protocol.ts [hunk 6]
```diff
@@ -507,32 +502,22 @@ export namespace UserEnvVar {
...
+            ...patternSegments.map((s) => {
+                switch (s) {
+                    case WILDCARD_ASTERISK:
+                        return "[^/]*";
+                    case WILDCARD_ASTERISK:
+                        return "[^/]*";
+                    case WILDCARD_DOUBLE_ASTERISK:
+                        return ".*";
+                    default:
+                        return s;
```
Comment (svenefftinge) [inline components/gitpod-protocol/src/protocol.ts:515]: asked chatGPT for replacing control patterns:
```
s.replace(/[-\/\\^$+?.()|[\]{}]/g, '\\$&')
```
didn't try 🤷

File: components/gitpod-protocol/src/protocol.ts [hunk 7]
```diff
@@ -466,55 +473,100 @@ export namespace UserEnvVar {
...
+                }
+            }),
+        );
+        const regex = new RegExp(regexStr);
```
Comment (svenefftinge) [inline components/gitpod-protocol/src/protocol.ts:519]: new RegExp('^' + regexStr + '$')

**Comment by svenefftinge** [comment]
Cool! I find the code much simpler tbh as we no longer have the iteration over two arrays logic

**Review by svenefftinge — APPROVED** [review]

**Pull request merged**